File: src/system_state.rs
use crate::battery::{
    ACPIType, BatteryStates, BatteryStatesError, ChargingStatus, PlatformProfile,
};
use crate::cpu::{CpuStates, CpuStatesError, CpuType, ScalingGoverner, EPP};
use std::cell::RefCell;
use std::fmt;
use std::fs;
use std::io;
use std::path::Path;

#[derive(Debug)]
pub enum SystemStateError {
    ACPITypeErr(String),

    CpuStatesErr(CpuStatesError),
    BatteryStatesErr(BatteryStatesError),
    GeneralIoErr(io::Error),
}

impl fmt::Display for SystemStateError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            SystemStateError::ACPITypeErr(e) => write!(f, "{e}"),
            SystemStateError::CpuStatesErr(e) => write!(f, "{e}"),
            SystemStateError::BatteryStatesErr(e) => write!(f, "{e}"),
            SystemStateError::GeneralIoErr(e) => write!(f, "General io error: {e}"),
        }
    }
}

impl From<CpuStatesError> for SystemStateError {
    fn from(error: CpuStatesError) -> Self {
        SystemStateError::CpuStatesErr(error)
    }
}

impl From<BatteryStatesError> for SystemStateError {
    fn from(error: BatteryStatesError) -> Self {
        SystemStateError::BatteryStatesErr(error)
    }
}

impl From<io::Error> for SystemStateError {
    fn from(error: io::Error) -> Self {
        SystemStateError::GeneralIoErr(error)
    }
}

#[derive(PartialEq, Debug, Clone, Copy)]
pub enum State {
    Powersave,
    Balanced,
    Performance,
}

pub struct SystemState {
    pub linux: bool,
    pub cpu_type: CpuType,
    pub acpi_type: ACPIType,

    pub cpu_states: CpuStates,
    pub battery_states: BatteryStates,

    pub state: RefCell<State>,
}

impl fmt::Display for SystemState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}\n{}\nSystem state: {:?}",
            self.cpu_states,
            self.battery_states,
            *self.state.borrow(),
        )
    }
}

impl SystemState {
    pub fn init() -> Result<Self, SystemStateError> {
        let cpu_type = Self::detect_cpu_type();
        let cpu_states = CpuStates::init(Self::num_cpu_cores()?, &cpu_type)?;
        Ok(Self {
            linux: Self::detect_linux(),
            cpu_type,
            acpi_type: Self::detect_acpi_type(),

            cpu_states,
            battery_states: BatteryStates::init()?,

            state: RefCell::new(State::Powersave),
        })
    }

    pub fn post_init(&self) -> Result<(), SystemStateError> {
        match self.battery_states.read_charging_status()? {
            ChargingStatus::Charging | ChargingStatus::NotCharging => {
                self.set_performance_mode(false)?;
                *self.state.borrow_mut() = State::Performance;
            }
            ChargingStatus::DisCharging => {
                self.set_powersave_mode()?;
                *self.state.borrow_mut() = State::Powersave;
            }
            ChargingStatus::Unknown => {
                self.set_balanced_mode()?;
                *self.state.borrow_mut() = State::Powersave;
            }
        }

        Ok(())
    }

    pub fn set_powersave_mode(&self) -> Result<(), SystemStateError> {
        self.cpu_states
            .set_scaling_governer(ScalingGoverner::Powersave)?;

        self.cpu_states.set_epp(EPP::Power)?;

        self.battery_states
            .set_platform_profile(&PlatformProfile::LowPower)?;

        self.cpu_states.set_cpu_boost(false)?;

        Ok(())
    }

    // for now, only for high cpu temp situations when charging
    pub fn set_balanced_mode(&self) -> Result<(), SystemStateError> {
        self.cpu_states
            .set_scaling_governer(ScalingGoverner::Powersave)?;

        self.cpu_states.set_epp(EPP::BalancePower)?;

        self.battery_states
            .set_platform_profile(&PlatformProfile::Balanced)?;

        self.cpu_states.set_cpu_boost(false)?;

        Ok(())
    }

    pub fn set_performance_mode(&self, cpu_boost: bool) -> Result<(), SystemStateError> {
        if self.battery_states.read_charging_status()? == ChargingStatus::DisCharging {
            return Ok(());
        }

        self.cpu_states
            .set_scaling_governer(ScalingGoverner::Performance)?;

        self.cpu_states.set_epp(EPP::Performance)?;

        self.battery_states
            .set_platform_profile(&PlatformProfile::Performance)?;

        self.cpu_states.set_cpu_boost(cpu_boost)?;

        Ok(())
    }

    // TODO: look over all of these again bc it's vibe coded
    fn detect_linux() -> bool {
        #[cfg(target_os = "linux")]
        let compile_time = true;
        #[cfg(not(target_os = "linux"))]
        let compile_time = false;

        let runtime_uname = std::process::Command::new("uname")
            .arg("-s")
            .output()
            .ok()
            .and_then(|o| String::from_utf8(o.stdout).ok())
            .map(|s| s.trim().eq_ignore_ascii_case("linux"))
            .unwrap_or(false);

        let has_proc = Path::new("/proc").exists();
        let has_sys = Path::new("/sys").exists();
        let has_etc = Path::new("/etc").exists();

        let has_os_release =
            Path::new("/etc/os-release").exists() || Path::new("/usr/lib/os-release").exists();

        compile_time
            || (runtime_uname && has_proc && has_sys)
            || (has_proc && has_sys && has_etc && has_os_release)
    }

    fn detect_cpu_type() -> CpuType {
        if let Ok(cpuinfo) = fs::read_to_string("/proc/cpuinfo") {
            for line in cpuinfo.lines() {
                if line.starts_with("vendor_id") {
                    if line.contains("GenuineIntel") {
                        return CpuType::Intel;
                    } else if line.contains("AuthenticAMD") {
                        return CpuType::AMD;
                    }
                }
            }
        }

        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
        {
            if let Some(cpu_type) = Self::detect_cpu_via_cpuid() {
                return cpu_type;
            }
        }

        if let Ok(output) = std::process::Command::new("lscpu").output() {
            if let Ok(text) = String::from_utf8(output.stdout) {
                let lower = text.to_lowercase();
                if lower.contains("genuineintel") || lower.contains("intel") {
                    return CpuType::Intel;
                } else if lower.contains("authenticamd") || lower.contains("amd") {
                    return CpuType::AMD;
                }
            }
        }

        CpuType::Unknown
    }

    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
    fn detect_cpu_via_cpuid() -> Option<CpuType> {
        #[cfg(target_arch = "x86")]
        use core::arch::x86::__cpuid;
        #[cfg(target_arch = "x86_64")]
        use core::arch::x86_64::__cpuid;

        unsafe {
            // CPUID leaf 0 returns vendor string in EBX, EDX, ECX
            let result = __cpuid(0);

            // convert registers to bytes
            let mut vendor = [0u8; 12];
            vendor[0..4].copy_from_slice(&result.ebx.to_le_bytes());
            vendor[4..8].copy_from_slice(&result.edx.to_le_bytes());
            vendor[8..12].copy_from_slice(&result.ecx.to_le_bytes());

            match &vendor {
                b"GenuineIntel" => Some(CpuType::Intel),
                b"AuthenticAMD" => Some(CpuType::AMD),
                _ => None,
            }
        }
    }

    fn num_cpu_cores() -> Result<usize, SystemStateError> {
        let cpu_dir = "/sys/devices/system/cpu/";
        let mut count = 0;

        for entry in fs::read_dir(cpu_dir)? {
            let entry = entry?;
            let path = entry.path();

            if path.is_dir() {
                if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
                    if name.starts_with("cpu") && name != "cpufreq" {
                        // Check if the remaining part of the name is a number
                        if name[3..].parse::<u32>().is_ok() {
                            count += 1;
                        }
                    }
                }
            }
        }
        Ok(count)
    }

    fn detect_acpi_type() -> ACPIType {
        if let Ok(product_version) = fs::read_to_string("/sys/class/dmi/id/product_version") {
            let product_version = product_version.trim().to_lowercase();
            if product_version.contains("thinkpad") {
                return ACPIType::ThinkPad;
            }
            //if product_version.contains("ideapad") {
            //    return ACPIType::IdeaPad;
            //}
        }

        if let Ok(product_name) = fs::read_to_string("/sys/class/dmi/id/product_name") {
            let product_name = product_name.trim().to_lowercase();
            if product_name.contains("thinkpad") {
                return ACPIType::ThinkPad;
            }
            //if product_name.contains("ideapad") {
            //    return ACPIType::IdeaPad;
            //}
        }

        if Path::new("/proc/acpi/ibm").exists() {
            return ACPIType::ThinkPad;
        }

        ACPIType::Unknown
    }
}

File: src/system_state.zig
const std = @import("std");
const builtin = @import("builtin");
const l_root = @import("root.zig");
const l_cpu_states = @import("cpu_states.zig");
const l_battery_states = @import("battery_states.zig");

const mem = std.mem;
const CpuStates = l_cpu_states.CpuStates;
const CpuType = l_cpu_states.CpuType;
const ScalingGoverner = l_cpu_states.ScalingGoverner;
const EPP = l_cpu_states.EPP;
const PlatformProfile = l_battery_states.PlatformProfile;
const BatteryStates = l_battery_states.BatteryStates;
const AcpiType = l_battery_states.AcpiType;
const ChargingStatus = l_battery_states.ChargingStatus;
const PersFd = l_root.PersFd;

pub const State = enum { Powersave, Balanced, Performance };
pub const SystemStateError = error{InvalidAcpiType};
pub const SystemState = struct {
    linux: bool,
    cpu_type: CpuType,
    acpi_type: AcpiType,

    cpu_states: CpuStates,
    battery_states: BatteryStates,

    state: State,

    pub fn init() !@This() {
        const cpu_type = SystemState.detectCpuType();
        return .{
            .linux = SystemState.detectLinux(),
            .cpu_type = cpu_type,
            .acpi_type = try SystemState.detectAcpiType(),
            .cpu_states = try CpuStates.init(cpu_type),
            .battery_states = try BatteryStates.init(),
            .state = .Powersave,
        };
    }

    pub fn postInit(self: *@This()) !void {
        const status = try self.battery_states.readChargingStatus();
        switch (status) {
            ChargingStatus.Charging, ChargingStatus.NotCharging => {
                try self.setPerformanceMode(false);
                self.state = .Performance;
            },
            ChargingStatus.DisCharging => {
                self.setPowersaveMode();
                self.state = .Powersave;
            },
            ChargingStatus.Unknown => {
                self.setPowersaveMode();
                self.state = .Powersave;
            },
        }
    }

    pub fn deinit(self: *@This()) void {
        self.battery_states.deinit();
        self.cpu_states.deinit();
    }

    pub fn print(self: *@This()) !void {
        try self.cpu_states.print();
        try self.battery_states.print();
        std.debug.print("State: {any}\n", .{self.state});
    }

    pub fn setPowersaveMode(self: *@This()) void {
        self.cpu_states.setScalingGoverner(ScalingGoverner.Powersave) catch |err| {
            std.log.warn("Failed to set scaling governor to Powersave: {}", .{err});
        };
        self.cpu_states.setEPP(EPP.Power) catch |err| {
            std.log.warn("Failed to set EPP to Power: {}", .{err});
        };
        self.battery_states.setPlatformProfile(PlatformProfile.LowPower) catch |err| {
            std.log.warn("Failed to set platform profile to LowPower: {}", .{err});
        };
        self.cpu_states.setCpuBoost(false) catch |err| {
            std.log.warn("Failed to disable CPU boost: {}", .{err});
        };
    }

    // for now, only for high cpu temp situations when charging
    pub fn setBalancedMode(self: *@This()) void {
        self.cpu_states.setScalingGoverner(ScalingGoverner.Powersave) catch |err| {
            std.log.warn("Failed to set scaling governor to Powersave: {}", .{err});
        };
        self.cpu_states.setEPP(EPP.BalancePower) catch |err| {
            std.log.warn("Failed to set EPP to BalancePower: {}", .{err});
        };
        self.battery_states.setPlatformProfile(PlatformProfile.Balanced) catch |err| {
            std.log.warn("Failed to set platform profile to Balanced: {}", .{err});
        };
        self.cpu_states.setCpuBoost(false) catch |err| {
            std.log.warn("Failed to disable CPU boost: {}", .{err});
        };
    }

    pub fn setPerformanceMode(self: *@This(), enable_boost: bool) !void {
        if (try self.battery_states.readChargingStatus() == ChargingStatus.DisCharging)
            return;

        self.cpu_states.setScalingGoverner(ScalingGoverner.Performance) catch |err| {
            std.log.warn("Failed to set scaling governor to Performance: {}", .{err});
        };
        self.cpu_states.setEPP(EPP.Performance) catch |err| {
            std.log.warn("Failed to set EPP to Performance: {}", .{err});
        };
        self.battery_states.setPlatformProfile(PlatformProfile.Performance) catch |err| {
            std.log.warn("Failed to set platform profile to Performance: {}", .{err});
        };
        self.cpu_states.setCpuBoost(enable_boost) catch |err| {
            std.log.warn("Failed to set CPU boost to {}: {}", .{enable_boost, err});
        };
    }

    fn detectLinux() bool {
        const compile_time = if (builtin.os.tag == .linux) true else false;
        const proc_exists = if (std.fs.cwd().access("/proc", .{})) true else |_| false;
        const sys_exists = if (std.fs.cwd().access("/sys", .{})) true else |_| false;
        const etc_exists = if (std.fs.cwd().access("/etc", .{})) true else |_| false;

        const etc_os_release = if (std.fs.cwd().access("/etc/os-release", .{})) true else |_| false;
        const usr_os_release = if (std.fs.cwd().access("/usr/lib/os-release", .{})) true else |_| false;
        const has_os_release = etc_os_release or usr_os_release;

        return compile_time or (proc_exists and sys_exists) or (proc_exists and sys_exists and etc_exists and has_os_release);
    }

    fn detectCpuType() CpuType {
        var fd = PersFd.init("/proc/cpuinfo", false) catch
            return CpuType.Unknown;
        defer fd.close();

        const val = fd.readValue() catch
            return CpuType.Unknown;

        var line_iter = mem.splitScalar(u8, val, '\n');
        while (line_iter.next()) |line| {
            if (mem.startsWith(u8, line, "vendor_id")) {
                if (mem.indexOf(u8, line, "GenuineIntel") != null) {
                    return .Intel;
                } else if (mem.indexOf(u8, line, "AuthenticAMD") != null) {
                    return .AMD;
                }
            }
        }

        if (detectCpuViaCpuid()) |cpu_type|
            return cpu_type;

        return CpuType.Unknown;
    }

    fn detectCpuViaCpuid() ?CpuType {
        if (builtin.cpu.arch != .x86 and builtin.cpu.arch != .x86_64)
            return null;

        var eax: u32 = undefined;
        var ebx: u32 = undefined;
        var ecx: u32 = undefined;
        var edx: u32 = undefined;

        // cpuid(eax=0) -> vendor id in ebx, ecx, edx
        asm volatile ("cpuid"
            : [_eax] "={eax}" (eax),
              [_ebx] "={ebx}" (ebx),
              [_ecx] "={ecx}" (ecx),
              [_edx] "={edx}" (edx),
            : [leaf] "{eax}" (@as(u32, 0)),
            : .{ .memory = true });

        var vendor: [12]u8 = undefined;
        @memcpy(vendor[0..4], mem.asBytes(&ebx));
        @memcpy(vendor[4..8], mem.asBytes(&edx));
        @memcpy(vendor[8..12], mem.asBytes(&ecx));

        if (mem.eql(u8, &vendor, "GenuineIntel")) {
            return CpuType.Intel;
        } else if (mem.eql(u8, &vendor, "AuthenticAMD")) {
            return CpuType.AMD;
        }

        return null;
    }

    fn detectAcpiType() !AcpiType {
        const thinkpad = "thinkpad";
        const ideapad = "ideapad";

        var pv = try PersFd.init("/sys/class/dmi/id/product_version", false);
        defer pv.close();
        if (pv.readValue()) |product_version| {
            const trimmed = mem.trim(u8, product_version, &std.ascii.whitespace);
            var lowered: [pv.buffer.len]u8 = undefined;
            _ = std.ascii.lowerString(&lowered, trimmed);

            if (mem.indexOf(u8, &lowered, thinkpad) != null) return AcpiType.ThinkPad;
            if (mem.indexOf(u8, &lowered, ideapad) != null) return AcpiType.IdeaPad;
        } else |_| {}

        var pn = try PersFd.init("/sys/class/dmi/id/product_name", false);
        defer pn.close();
        if (pn.readValue()) |product_name| {
            const trimmed = mem.trim(u8, product_name, &std.ascii.whitespace);
            var lowered: [pv.buffer.len]u8 = undefined;
            _ = std.ascii.lowerString(&lowered, trimmed);
            if (mem.indexOf(u8, &lowered, thinkpad) != null) return AcpiType.ThinkPad;
            if (mem.indexOf(u8, &lowered, ideapad) != null) return AcpiType.IdeaPad;
        } else |_| {}

        if (std.fs.cwd().access("/proc/acpi/ibm", .{})) {
            return AcpiType.ThinkPad;
        } else |_| {}

        return AcpiType.Unknown;
    }
};
